### PR TITLE
Adding New Modules

### PR DESCRIPTION
**Pull request by jbadger95**
Adding a few modules that I used to support some other changes.

- bitvisit -- Stores and manipulates array of visitation flags (optionally bitpacks flags) that is convenient for reducing across MPI ranks
- index_map -- Maps a collection of array items to their location (index) in the array
- sorts -- A collection of sorting routines used throughout the code

### CODE REVIEW

**Comment by stefanhenneking** [comment]
Are there any sorting routines that you copy-pasted from elsewhere in this new module and which can now be removed elsewhere?

File: trunk/src/modules/sorts.F90 [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+!
+!> Purpose - sorts an array of duplets (iel,residual)
+!! @param[inout] Iel_array - Index array for permutation
+!! @param[inout] Residuals - array to sort
+!! @param[in]    N         - list of integers to sort
+!! @param[in]    First     - first index of current partition (needs to be 1 on initial call)
+!! @param[in]    Last      - last index of current partition (needs to be N on initial call)
+   recursive subroutine qsort_duplet(Iel_array,Residuals,N,First,Last)
+!
+      implicit none
```
Comment (stefanhenneking) [inline trunk/src/modules/sorts.F90:164]:
```suggestion
```

File: trunk/src/modules/sorts.F90 [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+!!           IA (major), JA (minor), in ascending order
+!! @param[inout] ia    - integer array (row indices)
+!! @param[inout] ja    - integer array (colun indices)
+!! @param[inout] xa    - array of values
+!! @param[in]    n     - list of integers to sort
+!! @param[in]    first - first index of current partition (1 on initial call)
+!! @param[in]    last  - last index of current partition (N on initial call)
+   recursive subroutine qsort_triplet(ia,ja,xa,n,first,last)
+!
+      implicit none
```
Comment (stefanhenneking) [inline trunk/src/modules/sorts.F90:214]:
```suggestion
```

File: trunk/src/modules/sorts.F90 [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+!! @param[inout] ia    - integer array (row indices)
+!! @param[inout] ja    - integer array (colun indices)
+!! @param[inout] xa    - array of values
+!! @param[out]   k     - pivot
+!! @param[in]    n     - list of integers to sort
+!! @param[in]    first - first index of current partition (1 on initial call)
+!! @param[in]    last  - last index of current partition (N on initial call)
+   recursive subroutine partition_triplet_omp(ip,jp,n,k,ia,ja,xa)
+!
+      implicit none
```
Comment (stefanhenneking) [inline trunk/src/modules/sorts.F90:263]:
```suggestion
```

File: trunk/src/modules/sorts.F90 [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+end subroutine partition_triplet_omp
+
+
+
+
+
+!> Purpose - Threaded version of qsort_triplet
+recursive subroutine qsort_triplet_omp(n,ia,ja,xa)
+!
+   implicit none
```
Comment (stefanhenneking) [inline trunk/src/modules/sorts.F90:328]:
```suggestion
```

File: trunk/src/modules/sorts.F90 [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+
+
+
+
+!  Note: This array is similar to qsort_duplet but sorts by index, instead of value
+!
+!> Purpose - sorts duplet (ia,val) by index
+   recursive subroutine qsort_double(ia,val,n,first, last)
+!
+      implicit none
```
Comment (stefanhenneking) [inline trunk/src/modules/sorts.F90:386]:
```suggestion
```

**Review by stefanhenneking — APPROVED** [review]

**Comment by jbadger95** [comment]
I tried to copy all the sorting routines (don't know if I got all of them). I have not replaced them in the code yet though; to do this routines using sorts just need to use the `sorts` module and the routines need to be deleted; I can do this but I wasn't sure if you thought it would be more clear to have some of these functions defined where they are used (like in the sparse matrix routines). The main reason for committing this was that my `index_map` module needs one of the routines and I already had them all modularized. `index_map` is used in some uncommited changes for pre-computing element nodes (we decided we may not add that, I still may add the module after testing though; it doesn't have to be used if not wanted). It's also used heavily in my MG solver

**Comment by stefanhenneking** [comment]
I like the idea of having a sort module that's used consistently in the code and remove sorting functions elsewhere. But we can wait with the removal of other sorting functions until having merged #75 which removes some of them also to avoid the conflicts.

File: trunk/src/modules/sorts.F90 [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+
+
+!> Purpose - partitions triplet array around a pivot
+!! @param[inout] ia    - integer array (row indices)
+!! @param[inout] ja    - integer array (colun indices)
+!! @param[inout] xa    - array of values
+!! @param[out]   k     - pivot
+!! @param[in]    n     - list of integers to sort
+!! @param[in]    first - first index of current partition (1 on initial call)
+!! @param[in]    last  - last index of current partition (N on initial call)
```
Comment (stefanhenneking) [inline trunk/src/modules/sorts.F90:260]: update params

**Comment by stefanhenneking** [comment]
@jbadger95 can you check on the comment/suggestions. then, it's ready to be merged imo.

File: trunk/src/modules/bitvisit.F90 [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+      if (bitvisit_packed) then
+!     ...get word number for entry
+         iword = (i - 1)/bpw + 1
+!$omp critical
+         if (val .eq. 1) then
+            visitation(iword) = ibset(visitation(iword),mod(i-1,bpw))
+         else
+            visitation(iword) = ibclr(visitation(iword),mod(i-1,bpw))
+         endif
+!$omp end critical
```
[the remark this answers is not in the document]
Comment (jbadger95) [reply trunk/src/modules/bitvisit.F90:109]: visitation is the array that stores whether something has been visited (it is sized when the initialization routine is called). It's frequently of size NRNODS so no, threadprivate is not an option.

**Comment by jbadger95** [comment]
Ok, I addressed most of the comments. Removing redundant `implicit none`s is a minor stylistic difference (I like having it in the functions for clarity) so I think we can let these slide.

File: trunk/src/modules/bitvisit.F90 (hunk 7 above)
[the remark this answers is not in the document]
Comment (stefanhenneking) [reply trunk/src/modules/bitvisit.F90:109]: ok, maybe we can yet think of another way to "visit" more efficiently in the future (of course, if `NODES` were distributed, they issue becomes more or less irrelevant in terms of performance).

**Comment by stefanhenneking** [comment]
Sounds good, merge when you're ready.

**Pull request merged**